File: playwright-java/src/test/java/testtracks/steps/restfulbooker/api/auth/AuthApi.java
package testtracks.steps.restfulbooker.api.auth;

import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.microsoft.playwright.APIRequestContext;
import com.microsoft.playwright.APIResponse;
import java.util.Map;
import testtracks.api.common.BaseApi;
import testtracks.configuration.TestSettings;
import testtracks.diagnostics.ApiEvidence;

public final class AuthApi extends BaseApi {
  public AuthApi(TestSettings settings, APIRequestContext request, ApiEvidence evidence) {
    super(settings, request, evidence);
  }

  public AuthToken logIn(ObjectNode payload) {
    return logIn(payload, null, null);
  }

  public AuthToken logIn(
      ObjectNode payload, Map<String, String> headers, Map<String, String> query) {
    return sendJson("POST", "auth/login", payload, headers, query, AuthToken.class);
  }

  public APIResponse tryLogIn(ObjectNode payload) {
    return send("POST", "auth/login", payload, null, null);
  }

  public TokenValidation validate(String token) {
    return sendJson(
        "POST", "auth/validate", tokenPayload(token), null, null, TokenValidation.class);
  }

  public void logOut(String token) {
    ensureOk("POST", "auth/logout", tokenPayload(token), null, null);
  }

  private static ObjectNode tokenPayload(String token) {
    var payload = JsonNodeFactory.instance.objectNode();
    payload.put("token", token);
    return payload;
  }
}

File: playwright-csharp/tests/Steps/RestfulBooker/API/Auth/AuthApi.cs
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
using Microsoft.Playwright;
using TestTracks.Playwright.CSharp.API.Common;
using TestTracks.Playwright.CSharp.Configuration;
using TestTracks.Playwright.CSharp.Diagnostics;

namespace TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Auth;

public sealed class AuthApi : BaseApi
{
    public AuthApi(TestSettings settings, IAPIRequestContext request, ApiEvidence? evidence = null)
        : base(settings, request, evidence)
    {
    }

    public async Task<AuthToken> LogIn(
        JsonObject payload,
        IReadOnlyDictionary<string, string>? headers = null,
        IReadOnlyDictionary<string, string?>? query = null)
    {
        return await SendJsonAsync<AuthToken>("POST", "auth/login", payload, headers, query);
    }

    public async Task<IAPIResponse> TryLogIn(
        JsonObject payload,
        IReadOnlyDictionary<string, string>? headers = null,
        IReadOnlyDictionary<string, string?>? query = null)
    {
        return await SendAsync("POST", "auth/login", payload, headers, query);
    }

    public async Task<TokenValidation> Validate(
        string token,
        IReadOnlyDictionary<string, string>? headers = null,
        IReadOnlyDictionary<string, string?>? query = null)
    {
        return await SendJsonAsync<TokenValidation>("POST", "auth/validate", TokenPayload(token), headers, query);
    }

    public async Task LogOut(
        string token,
        IReadOnlyDictionary<string, string>? headers = null,
        IReadOnlyDictionary<string, string?>? query = null)
    {
        await EnsureOkAsync("POST", "auth/logout", TokenPayload(token), headers, query);
    }

    private static JsonObject TokenPayload(string token)
    {
        return new JsonObject
        {
            ["token"] = token
        };
    }
}

public sealed record AuthToken(
    [property: JsonPropertyName("token")] string Token);

public sealed record TokenValidation(
    [property: JsonPropertyName("valid")] bool Valid);
